GENOMES_DIR='/media/data1/genomes'
OUT_DIR = '/media/data2/HuR_human_mouse/mouse/ribo-seq'
SRC_DIR = '/home/saket/github_projects/clip_seq_pipeline/scripts'
RAWDATA_DIR ='/media/data1/HuR_human_mouse/mouse/ribo/Penalva_L_03222017'
GENOME_BUILD = 'mm10'
GENOME_FASTA = GENOMES_DIR + '/' + GENOME_BUILD + '/fasta/'+ GENOME_BUILD+ '.fa'
STAR_INDEX = GENOMES_DIR + '/' + GENOME_BUILD + '/star_annotated'
GTF = GENOMES_DIR + '/' + GENOME_BUILD + '/annotation/' + 'gencode.vM11.annotation.without_rRNA_tRNA.gtf'
GENE_NAMES = GENOMES_DIR + '/' + GENOME_BUILD + '/annotation/' + GENOME_BUILD+'_gene_names_stripped.tsv'
GENE_LENGTHS = GENOMES_DIR + '/' + GENOME_BUILD + '/annotation/' + 'gencode.vM11.coding_lengths.tsv'  #+ GENOME_BUILD+'_gene_lengths.tsv'
DESIGN_FILE = RAWDATA_DIR + '/' + 'design.txt'
HTSEQ_STRANDED = 'yes'
FEATURECOUNTS_S = '-s 1'
GENE_BED = GENOMES_DIR + '/' + GENOME_BUILD + '/annotation/' + 'gencode.vM11.genes.fromucsc.bed'  #+ GENOME_BUILD+'_gene_lengths.tsv'
FEATURECOUNTS_T='cds'
HTSEQ_MODE='union'
