include:
    '../configs/config_HuR_Penalva_L_01182017.human.rsem.py'
workdir: OUT_DIR

SAMPLE_NAMES_RNA = ['HepG2_Control_1_S7',
                    'HepG2_Control_2_S9',
                    'HepG2_Control_6_S11',
                    'HepG2_HuR_KD_1_S8',
                    'HepG2_HuR_KD2_S10',
                    'HepG2_HuR_KD_6_S12']
LANE_NAMES_RNA = ['L001', 'L002', 'L003', 'L004']
SAMPLE_NAMES_RIBO = ['HepG2_Control_1_S19',
                     'HepG2_Control_2_S21',
                     'HepG2_Control_6_S23',
                     'HepG2_HuR_KD_1_S20',
                     'HepG2_HuR_KD_2_S22',
                     'HepG2_HuR_KD_6_S24']

print(RAWDATA_DIR_RNA)

rule all:
    input:
        RSEM_INDEX_PREFIX + '.transcripts.fa',
        expand(RAWDATA_DIR_RNA+'/merged_fastq/{sample_name}_R1_001.fastq.gz', sample_name=SAMPLE_NAMES_RNA),
        expand('mapped/{sample_name}/{sample_name}.genes.results', sample_name=SAMPLE_NAMES_RNA),
        #expand('preprocessed/rna/{sample_name}/{sample_name}_{lane}_R1_001_val_1.fq.gz', zip, sample_name=SAMPLE_NAMES_RNA, lane=LANE_NAMES_RNA),
        #expand('preprocessed/rna/{sample_name}/{sample_name}_{lane}_R2_001_val_2.fq.gz', zip, sample_name=SAMPLE_NAMES_RNA, lane=LANE_NAMES_RNA),


rule prepare_reference:
    output: RSEM_INDEX_PREFIX + '.transcripts.fa'
    threads: 16
    shell:
        r'''
        rsem-prepare-reference --gtf {GTF} --star -p {threads} {GENOME_FASTA} {RSEM_INDEX_PREFIX}
        '''

rule merge_multi_lane_samples:
    ## Need to do this! :-/
    input:
        R1=expand('{raw_dir_rna}/{{sample_name}}_{lane}_R1_001.fastq.gz', raw_dir_rna=RAWDATA_DIR_RNA, lane=LANE_NAMES_RNA),
        R2=expand('{raw_dir_rna}/{{sample_name}}_{lane}_R2_001.fastq.gz', raw_dir_rna=RAWDATA_DIR_RNA, lane=LANE_NAMES_RNA)
    output:
        R1=RAWDATA_DIR_RNA+'/merged_fastq/{sample_name}_R1_001.fastq.gz',
        R2=RAWDATA_DIR_RNA+'/merged_fastq/{sample_name}_R2_001.fastq.gz'
    run:
        r1_files = ' '.join(input.R1)
        r2_files = ' '.join(input.R2)
        shell(r'''zcat {r1_files} > {output.R1}''')
        shell(r'''zcat {r2_files} > {output.R2}''')

rule perfom_trimming_rna:
    input:
        R1=RAWDATA_DIR_RNA+'/merged_fastq/{sample_name}_R1_001.fastq.gz',
        R2=RAWDATA_DIR_RNA+'/merged_fastq/{sample_name}_R2_001.fastq.gz'
    params:
        out_dir='preprocessed/rna/{sample_name}',
        phred_cutoff=5
    output:
        'preprocessed/rna/{sample_name}_R1_001_val_1.fq.gz',
        'preprocessed/rna/{sample_name}_R2_001_val_2.fq.gz',
    shell:
        r'''
            trim_galore --paired -o {params.out_dir} -q {params.phred_cutoff} {input.R1} {input.R2}
        '''

rule quantify:
    input:
        R1=RAWDATA_DIR_RNA+'/merged_fastq/{sample_name}_R1_001.fastq.gz',
        R2=RAWDATA_DIR_RNA+'/merged_fastq/{sample_name}_R2_001.fastq.gz'
    threads: 16
    params:
        prefix = 'mapped/{sample_name}/{sample_name}',
        tmpdir =  'mapped/tmp_dir/{sample_name}/'
    output: 'mapped/{sample_name}/{sample_name}.genes.results'
    shell:
        r'''mkdir -p {OUT_DIR}/{params.tmpdir} && rsem-calculate-expression --paired-end --star --star-output-genome-bam --estimate-rspd -p {threads} {input.R1} {input.R2} {RSEM_INDEX_PREFIX} {params.prefix} --temporary-folder {OUT_DIR}/{params.tmpdir}'''
